src/main/java/net/zhuoweizhang/raspberryjuice/RemoteSession.java

```diff
@@ -222,43 +283,123 @@ protected void handleCommand(String c, String[] args) {
+						LivingEntity shooter = (LivingEntity)arrow.getShooter();
+						if (shooter.getEntityId() == entityId)
+							iter.remove();
+					}
+				}
 				
 			// events.block.hits
 			} else if (c.equals("events.block.hits")) {
+				int entityId = -1;
+				if (args[0].length() > 0)
+					entityId = Integer.parseInt(args[0]);
 				StringBuilder b = new StringBuilder();
-		 		PlayerInteractEvent event;
-				while ((event = interactEventQueue.poll()) != null) {
-					Block block = event.getClickedBlock();
-					Location loc = block.getLocation();
-					b.append(blockLocationToRelative(loc));
-					b.append(",");
-					b.append(blockFaceToNotch(event.getBlockFace()));
-					b.append(",");
-					b.append(event.getPlayer().getEntityId());
-					if (interactEventQueue.size() > 0) {
+				for (Iterator<PlayerInteractEvent> iter = interactEventQueue.iterator(); iter.hasNext(); ) {
+					PlayerInteractEvent event = iter.next();
+					if (entityId == -1 || event.getPlayer().getEntityId() == entityId) {
+						Block block = event.getClickedBlock();
+						Location loc = block.getLocation();
+						b.append(blockLocationToRelative(loc));
+						b.append(",");
+						b.append(blockFaceToNotch(event.getBlockFace()));
+						b.append(",");
+						b.append(event.getPlayer().getEntityId());
 						b.append("|");
+						iter.remove();
 					}
 				}
+				if (b.length() > 0)
+					b.deleteCharAt(b.length() - 1);
 				send(b.toString());
-			
+
 			// events.chat.posts
 			} else if (c.equals("events.chat.posts")) {
+				int entityId = -1;
+				if (args[0].length() > 0)
+					entityId = Integer.parseInt(args[0]);
 				StringBuilder b = new StringBuilder();
-				AsyncPlayerChatEvent event;
-				while ((event = chatPostedQueue.poll()) != null) {
-					b.append(event.getPlayer().getEntityId());
-					b.append(",");
-					b.append(event.getMessage());
-					if (chatPostedQueue.size() > 0) {
+				for (Iterator<AsyncPlayerChatEvent> iter = chatPostedQueue.iterator(); iter.hasNext(); ) {
+					AsyncPlayerChatEvent event = iter.next();
+					if (entityId == -1 || event.getPlayer().getEntityId() == entityId) {
+						b.append(event.getPlayer().getEntityId());
+						b.append(",");
+						b.append(event.getMessage());
 						b.append("|");
+						iter.remove();
 					}
 				}
+				if (b.length() > 0)
+					b.deleteCharAt(b.length() - 1);
 				send(b.toString());
 				
+			// events.projectile.hits
+			} else if(c.equals("events.projectile.hits"))
+			{
+				int entityId = -1;
+				if (args[0].length() > 0)
+					entityId = Integer.parseInt(args[0]);
+				StringBuilder b = new StringBuilder();
+				for (Iterator<ProjectileHitEvent> iter = projectileHitQueue.iterator(); iter.hasNext(); ) {
+					ProjectileHitEvent event = iter.next();
+					Arrow arrow = (Arrow) event.getEntity();
+					LivingEntity shooter = (LivingEntity)arrow.getShooter();
+					if (entityId == -1 || shooter.getEntityId() == entityId) {
+						if (shooter instanceof Player) {
```

Comment:
Should we include all projectiles? Why only those from players?

Reply:
Question is do anybody will be interested in collecting arrows hit of any entity ? When you think in terms multiplayer game, which player should get and clear these events. From player point of view you can say - let's every player collect their events individually. Collecting events from any entities also may overflow queue quickly I think.
Because queueProjectileHitEvent(ProjectileHitEvent event) method already collects only player's events, this condition can be removed from above part of code.

Reply:
> Question is do anybody will be interested in collecting arrows hit of any entity ?

Im not sure, but as an API, it shouldn't be restrictive.